### PR TITLE
Gssapi fix

### PR DESCRIPTION
**Pull request by BobSilent**
Fix authenticate for GSSAPI Usage, issue was introduced with multiple ldap server handling: In case of GSSAPI Usage only check password with GSSAPI and not also with bind.

Also in case of GSSAPI Usage the InitialDirContext is created by GSSAPI and not via Bind

### CODE REVIEW

**Comment by simonbrandhof** [comment]
Hi BobSilent, thanks for this PR but please format the new code according to conventions.

File: src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java [hunk 1]
```diff
@@ -91,6 +104,41 @@ public InitialDirContext createUserContext(String principal, String credentials)
...
+		  
+		  initialDirContext = Subject.doAs(lc.getSubject(), new PrivilegedAction<InitialDirContext>() {
+			  @Override
+			  public InitialDirContext run(){
+				  // TODO Auto-generated method stub
+				  try {
+					  return new InitialLdapContext(getEnvironment(null, null, pooling), null);
+				  } catch (NamingException e) {
+					  // TODO Auto-generated catch block
+					  e.printStackTrace();
```
Comment (simonbrandhof) [inline src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java]: This error must be correctly handled

File: src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java [hunk 2]
```diff
@@ -91,6 +104,41 @@ public InitialDirContext createUserContext(String principal, String credentials)
...
+	  InitialDirContext initialDirContext = null;
+	  LoginContext lc = null;
+	  try {
+		  lc = new LoginContext(getClass().getName(), new CallbackHandlerImpl(principal, credentials));
+		  lc.login();
+		  
+		  initialDirContext = Subject.doAs(lc.getSubject(), new PrivilegedAction<InitialDirContext>() {
+			  @Override
+			  public InitialDirContext run(){
+				  // TODO Auto-generated method stub
```
Comment (simonbrandhof) [inline src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java]: The comment should be deleted if irrelevant.

File: src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java [hunk 3]
```diff
@@ -91,6 +104,41 @@ public InitialDirContext createUserContext(String principal, String credentials)
     return createInitialDirContext(principal, credentials, false);
   }
 
+  private InitialDirContext createInitialDirContextUsingGssapi(String principal, String credentials, final boolean pooling) throws NamingException {
+	  Configuration.setConfiguration(new Krb5LoginConfiguration());
+	  //Configuration.getConfiguration();
+	  InitialDirContext initialDirContext = null;
+	  LoginContext lc = null;
+	  try {
+		  lc = new LoginContext(getClass().getName(), new CallbackHandlerImpl(principal, credentials));
```
Comment (simonbrandhof) [inline src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java]: Tabulations must be replaced by 2-whitespaces.

File: src/main/java/org/sonar/plugins/ldap/LdapAuthenticator.java [hunk 4]
```diff
@@ -95,6 +96,7 @@ private boolean checkPasswordUsingBind(String principal, String password, String
     InitialDirContext context = null;
     try {
       context = contextFactories.get(ldapKey).createUserContext(principal, password);
+	  LOG.debug("Password valid for user {} in server {}!", new Object[] {principal, ldapKey});
```
Comment (simonbrandhof) [inline src/main/java/org/sonar/plugins/ldap/LdapAuthenticator.java]: Master branch uses now SonarQube Loggers, so this statement should be replaced:
`LOG.debug("Password valid for user {} in server {}!", principal, ldapKey);`

File: src/main/java/org/sonar/plugins/ldap/LdapGroupMapping.java [hunk 5]
```diff
@@ -83,6 +87,16 @@ public LdapSearch createSearch(LdapContextFactory contextFactory, SearchResult u
...
+      } else if ("objectsid".equals(attr.toLowerCase())) {
+        Attribute attribute = user.getAttributes().get(attr);
+        byte[] objectSid;
+        try {
+          objectSid = (byte[])attribute.get();
+          PSID sid = new PSID(objectSid);
+          parameters[i] = Advapi32Util.convertSidToStringSid(sid);
+        } catch (NamingException e) {
+          parameters[i] = null;
+        }
```
Comment (Godin) [inline src/main/java/org/sonar/plugins/ldap/LdapGroupMapping.java]: Wondering how this relates to GSSAPI ?

**Pull request closed without merging**